src/protocol/batch.rs

```diff
@@ -0,0 +1,75 @@
+use std::ops::{Deref, DerefMut, Index, IndexMut};
+
+pub type RecordIndex = u32;
+
+/// A wrapper around Vec<T> ensuring that the wrapped vector has max length <= `RecordIndex::MAX`
+#[derive(Debug, Clone)]
+pub struct Batch<T>(Vec<T>);
```

Comment:
This wrapper ensures that the length of the inner vector is less than RecordIndex::MAX upon index references, but doesn't prevent it from growing larger at the time items are added (e.g., push(), append()). Is there a way to monitor changes in its size? @martinthomson @akoshelev

Reply:
It might be better to build this without exposing the internal collection, so that you only need to cover a few different interfaces for manipulation.

I imagine that you want to support `Iterator::collect()`, but maybe you don't need push, extend_from_slice() and so forth.